### PR TITLE
Provide Decoding Response example for Java #760

### PR DESCRIPTION
**Pull request by jlprat**
Issue: #760
Provide an example to show how to decode a response using the Java DSL
Add missing method in `HttpResponse` to retrieve he encoding

### CODE REVIEW

**Comment by jlprat** [comment]
I'm not so sure, if this is the best way to decode a response, the Java version is pretty verbose compared to the Scala one

**Comment by jlprat** [comment]
PLS BUILD

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by jlprat** [comment]
Ready for review

File: docs/src/test/java/docs/http/javadsl/HttpClientDecodingExample.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class HttpClientDecodingExample {
+
+  public static void main(String[] args) throws Exception {
+
+    final ActorSystem system = akka.actor.ActorSystem.create();
```
Comment (jonas) [inline docs/src/test/java/docs/http/javadsl/HttpClientDecodingExample.java]: Any reason this is using fully qualified class name?

Comment (jlprat) [inline docs/src/test/java/docs/http/javadsl/HttpClientDecodingExample.java]: Absolutely not,  I'll change it

**Comment by jlprat** [comment]
I think it's quite nice to have both APIs at the same level, so yes, I'm in. Should I do it in this same PR?

**Comment by jrudolph** [comment]
Do it as you like :)

**Comment by jlprat** [comment]
I'll do a deep dive into the transformations part in order to provide a solution for this, so it might take a little longer

**Comment by jlprat** [comment]
I think I'm missing one tiny bit in here. I'll write down what puzzles me.

AFAIK, the purpose of `HttpMessage.MessageTransformations.transformEntityDataBytes` would be to delegate the Data bytes transformation to its `HttpEntity`. This method should take a `Flow`, my first question would be: which one (Java or Scala DSL)? Any choice feels wrong to me as this method will end up in both `HttpRequest` and `HttpResponse` for both APIs.

Once both `HttpRequest` and `HttpResponse` end up having the `transformEntityDataBytes` method, I guess `DataMapper` is no longer needed for the Scala `Coder`, right?

Thanks in advance!

**Comment by jrudolph** [comment]
> AFAIK, the purpose of HttpMessage.MessageTransformations.transformEntityDataBytes would be to delegate the Data bytes transformation to its HttpEntity. This method should take a Flow, my first question would be: which one (Java or Scala DSL)? Any choice feels wrong to me as this method will end up in both HttpRequest and HttpResponse for both APIs.

You can use `Graph[FlowShape[In, Out], Mat]` which is extended by both Scala and Java `Flow`.

> Once both HttpRequest and HttpResponse end up having the transformEntityDataBytes method, I guess DataMapper is no longer needed for the Scala Coder, right?

We don't need it for `Decoder.decode` but we still need it for `Decoder.decodeData` if someone wants to run it on an `HttpEntity` directly. Not sure if that's an important use case but we would have to think about compatibility if we want to remove `decodeData`.

**Comment by jlprat** [comment]
Forgot to update MiMa filters and now I see this PR has conflicts, I will rebase

File: docs/src/test/scala/docs/http/scaladsl/HttpClientDecodingExampleSpec.scala [hunk 2]
```diff
@@ -47,6 +47,8 @@ class HttpClientDecodingExampleSpec extends WordSpec with Matchers with CompileO
 
     val futureResponses: Future[Seq[HttpResponse]] =
       Future.traverse(requests)(http.singleRequest(_).map(decodeResponse))
+
+    system.terminate()
```
Comment (jlprat) [inline docs/src/test/scala/docs/http/scaladsl/HttpClientDecodingExampleSpec.scala]: Scala example was not terminating the system making the example hang

File: akka-http/src/main/java/akka/http/javadsl/coding/Coder.java [hunk 3]
```diff
@@ -25,9 +29,40 @@
         this.underlying = underlying;
     }
 
+    public HttpResponse encodeMessage(HttpResponse message) {
```
Comment (jlprat) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: My Java generic typing Foo could not make it work in a better way, like for example expressing a `T` that extends both `HttpMessage` and `MessageTransformations`.
Comment is valid for the other `encodeMessage` and `decodeMessage` methods.

File: akka-http/src/main/java/akka/http/javadsl/coding/Coder.java [hunk 4]
```diff
@@ -25,9 +29,40 @@
...
+        boolean shouldApply = (boolean) underlying.messageFilter().apply((HttpMessage) message);
```
Comment (jlprat) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: `JavaMappings` is not so convenient to be used in Java code, so I fell back to the un-fancy cast.

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala [hunk 5]
```diff
@@ -417,7 +421,9 @@ final class HttpResponse(
 
   def mapEntity(f: ResponseEntity ⇒ ResponseEntity): HttpResponse = withEntity(f(entity))
 
-  /* Manual Case Class things, to easen bin-compat */
+  def transformEntityDataBytes[T](transformer: Graph[FlowShape[ByteString, ByteString], T]): HttpResponse = copy(entity = entity.transformDataBytes(Flow.fromGraph(transformer)))
```
Comment (jlprat) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: It's a `T` because some may pass a `Materializer`, others maybe not.

**Comment by jlprat** [comment]
PLS BUID

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala (hunk 5 above)
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: Might also work with just `Any`.

File: akka-http/src/main/mima-filters/10.0.3.backwards.excludes [hunk 6]
```diff
@@ -13,3 +13,7 @@ ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.marshal
 
 # make httpApp abstract class
 ProblemFilters.exclude[IncompatibleTemplateDefProblem]("akka.http.scaladsl.server.HttpApp")
+
+
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.coding.Encoder.encodeMessage")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.scaladsl.coding.Decoder.decodeMessage")
```
Comment (jrudolph) [inline akka-http/src/main/mima-filters/10.0.3.backwards.excludes]: These ones shouldn't be needed. Only the 10.0.4 ones should be needed. MiMa aggregates them for previous versions.

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala (hunk 5 above)
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: I see if that's also used in Java it would be problem with `Any` as Java doesn't respect or know about variance declarations.

File: akka-http/src/main/java/akka/http/javadsl/coding/Coder.java (hunk 3 above)
Comment (jrudolph) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: Hmm, might just not work. But let's not reimplement any stuff here. Why not just `(HttpResponse) underlying.encodeMessage(message)`?

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala [hunk 7]
```diff
@@ -5,18 +5,24 @@
...
 import headers.HttpEncoding
-import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.stream.scaladsl.{ Flow, Sink, Source }
 
 import scala.concurrent.Future
 
 trait Decoder {
   def encoding: HttpEncoding
 
+  def decodeMessage[T <: HttpMessage with MessageTransformations[T]](message: T): T#Self =
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: Could this be `def decodeMessage(message: HttpMessage): message.Self`? And please no copy pasting ;) Just delegate to the below method. We might need to introduce another `DataMapper[HttpMessage]`.

File: docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
+/*
```
Comment (jrudolph) [inline docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java]: Nice example.

File: akka-http/src/main/java/akka/http/javadsl/coding/Coder.java (hunk 3 above)
Comment (jlprat) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: This will not work, because `underlying.encodeMessage` needs a type that extends both `HttpMessage` and `MessageTranformations`. So what I could do is something like: `(HttpResponse) underlying.encodeMessage((SHttpResponse) response)`
Where `SHttpResponse` is the Scala one.

File: akka-http/src/main/mima-filters/10.0.3.backwards.excludes (hunk 6 above)
Comment (jlprat) [inline akka-http/src/main/mima-filters/10.0.3.backwards.excludes]: This PR started long time ago, and it was not in sync. I will remove them

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala (hunk 7 above)
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: OK

**Comment by jlprat** [comment]
I will address the comments and add a test for the Java functionality. Thanks!

File: akka-http/src/main/java/akka/http/javadsl/coding/Coder.java (hunk 3 above)
Comment (jrudolph) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: Ah, I see. You can add `def transformEntityDataBytes[T](transformer: Graph[FlowShape[ByteString, ByteString], T]): Self` to `scaladsl.HttpMessage` which will be implemented by the concrete implementations in `HttpRequest` and `HttpResponse`. Then

Comment (jlprat) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: Cool, I wasn't sure to add this method, then I can simplify the signature of the Scala's `encodeMessage` and `decodeMessage` to just take `HttpMessage` (instead of `HttpMessage with MessageTransformers`).
Then, I also question if the `transformEntityDataBytes` should be in `MessageTransformers` at all.

**Comment by jrudolph** [comment]
> Then, I also question if the transformEntityDataBytes should be in MessageTransformers at all.

That's the way how it gets into the Java API.

**Comment by jlprat** [comment]
> That's the way how it gets into the Java API.

Ah, OK I was speaking from the top of my head.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala (hunk 7 above)
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: Shouldn't be the other way round? `decode` calling `decodeMessage`, otherwise what is the purpose of the `transformEntityDataBytes` method?

File: akka-http/src/main/mima-filters/10.0.3.backwards.excludes (hunk 6 above)
Comment (jlprat) [inline akka-http/src/main/mima-filters/10.0.3.backwards.excludes]: Apparently, they are needed

File: akka-http/src/main/java/akka/http/javadsl/coding/Coder.java (hunk 3 above)
Comment (jlprat) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: Now they rely on the underlying encoder/decoder

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala [hunk 9]
```diff
@@ -10,18 +10,21 @@ import akka.stream.{ FlowShape, Materializer }
...
-import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.stream.scaladsl.{ Flow, Sink, Source }
 
 import scala.concurrent.Future
 
 trait Decoder {
   def encoding: HttpEncoding
 
-  def decode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
+  def decodeMessage(message: HttpMessage): HttpMessage#Self =
```
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: Keeping one implementation `decodeMessage` and having the old one calling this.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala [hunk 10]
```diff
@@ -10,18 +10,21 @@ import akka.stream.{ FlowShape, Materializer }
...
     if (message.headers exists Encoder.isContentEncodingHeader)
-      decodeData(message).withHeaders(message.headers filterNot Encoder.isContentEncodingHeader)
+      message.transformEntityDataBytes(decoderFlow).withHeaders(message.headers filterNot Encoder.isContentEncodingHeader)
     else message.self
 
+  def decode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
+    decodeMessage(message).asInstanceOf[T#Self]
```
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: I didn't know how else I could make the compiler happy

**Comment by jlprat** [comment]
Regarding the tests for the Java functionality, I could write them, but the current Java implementation is just delegating to the Scala implementation. Should I still add them?

File: akka-http-core/src/main/mima-filters/10.0.0.backwards.excludes [hunk 11]
```diff
@@ -13,3 +13,4 @@ ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.model.Ur
 ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.model.Uri.getUserInfo")
 ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.model.Uri.getPort")
 ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.model.Uri.getScheme")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.http.javadsl.model.HttpMessage#MessageTransformations.transformEntityDataBytes")
```
Comment (jrudolph) [inline akka-http-core/src/main/mima-filters/10.0.0.backwards.excludes]: Please don't add anything to old mima-filters. Adding to the 10.0.5 version should be enough.

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala [hunk 12]
```diff
@@ -140,6 +142,8 @@ sealed trait HttpMessage extends jm.HttpMessage {
   def withEntity(contentType: jm.ContentType, file: File): Self = withEntity(HttpEntity(contentType.asInstanceOf[ContentType], file))
   def withEntity(contentType: jm.ContentType, file: Path): Self = withEntity(HttpEntity.fromPath(contentType.asInstanceOf[ContentType], file))
 
+  def transformEntityDataBytes[T](transformer: Graph[FlowShape[ByteString, ByteString], T]): Self
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: :+1:

File: akka-http/src/main/java/akka/http/javadsl/coding/Coder.java (hunk 3 above)
Comment (jrudolph) [inline akka-http/src/main/java/akka/http/javadsl/coding/Coder.java]: :+1:

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala (hunk 9 above)
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: I think this can still be improved. We need to fix the `Self` type in `HttpMessage` to make it work (props to @rkuhn who figured out how to tie the recursive self-type knot for `FlowOps` a while ago).

If we change the `Self` type in `HttpMessage` to this:

```scala
type Self <: HttpMessage { type Self = HttpMessage.this.Self }
```

the type signature for `decodeMessage` can become

```scala
def decodeMessage(message: HttpMessage): message.Self
```

which means that whatever `HttpMessage` type is passed in will also come out.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala (hunk 10 above)
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: See above ;)

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala [hunk 13]
```diff
@@ -10,18 +10,21 @@ import akka.stream.{ FlowShape, Materializer }
...
   def encoding: HttpEncoding
 
-  def decode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
+  def decodeMessage(message: HttpMessage): HttpMessage#Self =
     if (message.headers exists Encoder.isContentEncodingHeader)
-      decodeData(message).withHeaders(message.headers filterNot Encoder.isContentEncodingHeader)
+      message.transformEntityDataBytes(decoderFlow).withHeaders(message.headers filterNot Encoder.isContentEncodingHeader)
     else message.self
 
+  def decode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: I think we can now deprecate this old type signature in favor of `decodeMessage`. Could you add the deprecation annotation?

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Encoder.scala [hunk 14]
```diff
@@ -18,11 +18,14 @@ trait Encoder {
 
   def messageFilter: HttpMessage ⇒ Boolean
 
-  def encode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Encoder.scala]: Same changes as for `Decoder`.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala (hunk 9 above)
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: Kudos to @rkuhn!! I will try this later and try to understand the trickery behind it :)

File: akka-http-core/src/main/mima-filters/10.0.0.backwards.excludes (hunk 11 above)
Comment (jlprat) [inline akka-http-core/src/main/mima-filters/10.0.0.backwards.excludes]: Actually it was complaining that needed to be added to this file, I don't know why. Maybe it's because this PR started long time ago and mima got confused...

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala (hunk 13 above)
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: Agree, didn't want to deprecate until you agreed on the solution

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala (hunk 9 above)
Comment (rkuhn) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: Glad I could help! :-)

**Comment by jlprat** [comment]
WOPS rebase went wrong...

**Comment by jlprat** [comment]
Ready to review.
As `decoder.decode` and `encoder.encode` are now deprecated, I guess we should go through the docs and update any occurrence with the new alternatives. I could do it here or in a different PR.

File: docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    final Function<HttpResponse, HttpResponse> decodeResponse = response -> {
+      // Pick the right coder
+      final Coder coder;
+      if (HttpEncodings.gzip().equals(response.encoding())) {
+        coder = Coder.Gzip;
+      } else if (HttpEncodings.deflate().equals(response.encoding())) {
+        coder = Coder.Deflate;
+      } else {
+        coder = Coder.NoCoding;
+      }
```
Comment (2beaucoup) [inline docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java]: We could do something like
```java
final List<Coder> decoders = Arrays.asList(GZip, Deflate, NoCoding)
final Coder = decoders.filter(d -> d.encoding.equals(response.encoding)).findFirst()
```
work here (same for Scala)?

File: akka-http/src/main/mima-filters/10.0.4.backwards.excludes [hunk 16]
```diff
@@ -0,0 +1,3 @@
+# Provide Decoding Response example for Java #760
```
Comment (jrudolph) [inline akka-http/src/main/mima-filters/10.0.4.backwards.excludes]: File can be removed.

File: docs/src/test/scala/docs/http/scaladsl/HttpClientDecodingExampleSpec.scala (hunk 2 above)
Comment (jlprat) [inline docs/src/test/scala/docs/http/scaladsl/HttpClientDecodingExampleSpec.scala]: I didn't look much on the Scala example as it was already there and approved. I agree with you, maybe it makes sense to collect the answers and log them.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala [hunk 17]
```diff
@@ -10,18 +10,21 @@ import akka.stream.{ FlowShape, Materializer }
...
   def encoding: HttpEncoding
 
-  def decode[T <: HttpMessage](message: T)(implicit mapper: DataMapper[T]): T#Self =
+  def decodeMessage(message: HttpMessage): message.Self =
     if (message.headers exists Encoder.isContentEncodingHeader)
-      decodeData(message).withHeaders(message.headers filterNot Encoder.isContentEncodingHeader)
+      message.transformEntityDataBytes(decoderFlow).withHeaders(message.headers filterNot Encoder.isContentEncodingHeader)
     else message.self
 
+  @deprecated("Use Decoder#decodeMessage instead. No need for implicit mapper.", since = "10.0.6")
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala]: :+1:

File: docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java (hunk 15 above)
Comment (jlprat) [inline docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java]: It is definitely more concise, and I guess it's still clear enough for the Java mainstream reader.

File: akka-http/src/main/mima-filters/10.0.4.backwards.excludes (hunk 16 above)
Comment (jlprat) [inline akka-http/src/main/mima-filters/10.0.4.backwards.excludes]: I knew I would forget to clean up at least one Mima file!

File: docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java (hunk 15 above)
Comment (jlprat) [inline docs/src/test/java/docs/http/javadsl/HttpClientDecodingExampleTest.java]: Not so easy for the Java code as `Coder` doesn't have access to `encoding`. Maybe this could be part of another PR.

**Comment by jlprat** [comment]
I just realized I forgot to update all references to the deprecated `encode` and `decode` to the new `encodeMessage` and `decodeMessage`.

**Comment by jlprat** [comment]
Failures are: #299 and #595 
PLS BUILD

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala (hunk 5 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: Yeah, so the `T` is very good here

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala [hunk 18]
```diff
@@ -274,7 +274,7 @@ final class HttpRequest(
   override def withUri(path: String): HttpRequest = withUri(Uri(path))
   def withUri(uri: Uri): HttpRequest = copy(uri = uri)
 
-  def transformEntityDataBytes[T](transformer: Graph[FlowShape[ByteString, ByteString], T]): HttpRequest = copy(entity = entity.transformDataBytes(Flow.fromGraph(transformer)))
+  def transformEntityDataBytes[M](transformer: Graph[FlowShape[ByteString, ByteString], M]): HttpRequest = copy(entity = entity.transformDataBytes(Flow.fromGraph(transformer)))
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: made it `M`, like we do for Materialized values everywhere

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala [hunk 19]
```diff
@@ -369,7 +375,7 @@ object HttpRequest {
       }
     }
 
-  /* Manual Case Class things, to easen bin-compat */
+  /* Manual Case Class things, to ease bin-compat */
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: thanks :)

**Comment by jlprat** [comment]
Yeah it was way more than expected! :smiley:
It started as a simple Scala to Java example conversion
